### ISSUE TITLE
spack won't use graphviz@2.30.1 from RHEL7 installation

### ISSUE THREAD

**Issue by goxberry** [spack/spack#8046]
Summary

The LLNL cluster `rztopaz` runs RHEL 7 and includes an installation of graphviz 2.30.1 that spack can't use without throwing an error. This error motivated PR #7962.

### Expected Result

The command `spack install librom` should install without errors.

### Actual Result

```console
[oxberry1@rztopaz572:spack]$ spack find
==> 8 installed packages.
-- linux-rhel7-x86_64 / intel@18.0.1 ----------------------------
cmake@2.8.12.2  elemental@0.87.7  intel-mkl@2018.0  metis@5.1.0  mvapich2@2.2  mvapich2@2.2  mvapich2@2.2  openblas@0.2.20
[oxberry1@rztopaz572:spack]$ spack install librom
==> boost@1.66.0 : externally installed in /usr/tce/packages/boost/boost-1.66.0-mvapich2-2.2-intel-18.0.1
==> boost@1.66.0 : generating module file
==> boost@1.66.0 : registering into DB
==> doxygen@1.8.5 : externally installed in /usr
==> doxygen@1.8.5 : generating module file
==> doxygen@1.8.5 : registering into DB
==> Error: Class constructor failed for package 'builtin.graphviz'.

Caused by:
NoURLError: Package Graphviz has no version with a URL.
  File "/collab/usr/gapps/librom/spack/lib/spack/spack/repository.py", line 833, in get
    return package_class(spec)
  File "/collab/usr/gapps/librom/spack/lib/spack/spack/package.py", line 616, in __init__
    f = fs.for_package_version(self, self.version)
  File "/collab/usr/gapps/librom/spack/lib/spack/spack/fetch_strategy.py", line 977, in for_package_version
    url = pkg.url_for_version(version)
  File "/collab/usr/gapps/librom/spack/lib/spack/spack/package.py", line 745, in url_for_version
    raise NoURLError(cls)

[oxberry1@rztopaz572:spack]$ spack install librom
==> boost@1.66.0 : externally installed in /usr/tce/packages/boost/boost-1.66.0-mvapich2-2.2-intel-18.0.1
==> boost@1.66.0 : already registered in DB
==> doxygen@1.8.5 : externally installed in /usr
==> doxygen@1.8.5 : already registered in DB
==> Error: Class constructor failed for package 'builtin.graphviz'.

Caused by:
NoURLError: Package Graphviz has no version with a URL.
  File "/collab/usr/gapps/librom/spack/lib/spack/spack/repository.py", line 833, in get
    return package_class(spec)
  File "/collab/usr/gapps/librom/spack/lib/spack/spack/package.py", line 616, in __init__
    f = fs.for_package_version(self, self.version)
  File "/collab/usr/gapps/librom/spack/lib/spack/spack/fetch_strategy.py", line 977, in for_package_version
    url = pkg.url_for_version(version)
  File "/collab/usr/gapps/librom/spack/lib/spack/spack/package.py", line 745, in url_for_version
    raise NoURLError(cls)
```

### Steps to reproduce the issue

```console
$ spack install librom
```

### Information on your system

I'm running spack on `rztopaz`, which should be `arch=linux-rhel7-x86_64`; I'm using `intel@18.0.1` to compile packages.

My `packages.yaml` is a bit verbose:

```console
packages:
  all:
    compiler: [intel@18.0.1, intel, gcc, pgi, clang, xl, nag]
    providers:
      mpi: [mvapich2]
      blas: [openblas, netlib-lapack, intel-mkl]
      lapack: [openblas, netlib-lapack, intel-mkl]
  # openblas:
  #   paths:
  #     openblas@0.2.20: /usr/
  #   buildable: False
  netlib-lapack:
    paths:
      netlib-lapack@3.4.2: /usr/
    buildable: False
  intel-mkl:
    paths:
      intel-mkl@2018.0: /usr/tce/packages/mkl/mkl-2018.0/
    buildable: False
  mvapich2:
    paths:
      mvapich2@2.2%intel@18.0.1: /usr/tce/packages/mvapich2/mvapich2-2.2-intel-18.0.1/
    buildable: False
  mpich:
    buildable: False
  openmpi:
    buildable: False
  intel-mpi:
    buildable: False
  spectrum-mpi:
    buildable: False
  intel-parallel-studio:
    buildable: False
  cmake:
    paths:
      cmake@2.8.12.2: /usr/
    buildable: False
  pkg-config:
    paths:
      pkg-config@0.27.1 arch=linux-rhel7-x86_64: /usr/bin/
    buildable: False
  m4:
    paths:
      m4@1.4.16 arch=linux-rhe17-x86_64: /usr/bin/
    buildable: False
  graphviz:
    paths:
      graphviz@2.30.1: /usr/
    buildable: False
  perl:
    paths:
      perl@5.16.3: /usr/
    buildable: False
  hdf5:
    paths:
      hdf5@1.8.18 %intel@18.0.1 ~mpi +cxx +fortran +hl +shared: /usr/tce/packages/hdf5/hdf5-serial-1.8.18-intel-18.0.1/
      hdf5@1.8.18 %intel@18.0.1 +mpi ~cxx +fortran +hl +shared ^mvapich2@2.2: /usr/tce/packages/hdf5/hdf5-parallel-1.8.18-intel-18.0.1-mvapich2-2.2/
    buildable: False
  bison:
    paths:
      bison@3.0.4: /usr/
    buildable: False
  boost:
    paths:
      boost@1.66.0 %intel@18.0.1 +atomic +chrono ~clanglibcpp +date_time +exception +filesystem +graph ~icu +iostreams +locale +log +math +mpi +program_options +python +regex +serialization +shared +signals +system  +test +thread +timer +wave ^mvapich@2.2: /usr/tce/packages/b\
oost/boost-1.66.0-mvapich2-2.2-intel-18.0.1/
    buildable: False
  doxygen:
    paths:
      doxygen@1.8.5: /usr/
    buildable: False
  netcdf:
    paths:
      netcdf@4.4.1.1 %intel@18.0.1: /usr/tce/packages/netcdf/netcdf-4.4.1.1-intel-18.0.1/
    buildable: False
  bzip2:
    paths:
      bzip2@1.0.6: /usr
    buildable: False
  python:
    paths:
      python@2.7.14: /usr/tce/
    buildable: False
  xz:
    paths:
      xz@5.2.2: /usr/
    buildable: False
  zlib:
    paths:
      zlib@1.2.7: /usr/
    buildable: False
  sundials:
    variants: +lapack
```

**Comment by citibeth** [comment]
This is good.  My guess is the problem will be solved by fixing the bug uncovered here.

**Comment by adamjstewart** [comment]
Thanks for the detailed bug report! For anyone who tries to debug this, a shorter means of reproducing the bug is the following `packages.yaml`:
```yaml
packages:
    graphviz:
        paths:
            graphviz@2.40.0: /usr/local
        buildable: False
```
```console
$ spack install comd -mpi+graphs
==> Error: Class constructor failed for package 'builtin.graphviz'.

Caused by:
NoURLError: Package Graphviz has no version with a URL.
  File "/Users/Adam/spack/lib/spack/spack/repository.py", line 833, in get
    return package_class(spec)
  File "/Users/Adam/spack/lib/spack/spack/package.py", line 616, in __init__
    f = fs.for_package_version(self, self.version)
  File "/Users/Adam/spack/lib/spack/spack/fetch_strategy.py", line 977, in for_package_version
    url = pkg.url_for_version(version)
  File "/Users/Adam/spack/lib/spack/spack/package.py", line 745, in url_for_version
    raise NoURLError(cls)
```
`comd` doesn't have any other dependencies, so it's pretty quick

**Comment by adamjstewart** [comment]
Pinging @alalazo

**Comment by adamjstewart** [comment]
Another user reported this same problem with the `intel-mkl` package on Slack, although the error message is very similar. To reproduce the problem, add `intel-mkl` as an external package and choose any version not listed in the package. Then, `spack spec intel-mkl` will fail with the following error message:
```
==> Error: Class constructor failed for package 'intel-mkl'.

Caused by:
TypeError: expected string or bytes-like object
  File "/Users/Adam/spack/lib/spack/spack/repo.py", line 833, in get
    return package_class(spec)
  File "/Users/Adam/spack/lib/spack/spack/package.py", line 672, in __init__
    f = fs.for_package_version(self, self.version)
  File "/Users/Adam/spack/lib/spack/spack/fetch_strategy.py", line 977, in for_package_version
    url = pkg.url_for_version(version)
  File "/Users/Adam/spack/lib/spack/spack/package.py", line 794, in url_for_version
    default_url, self.url_version(version))
  File "/Users/Adam/spack/lib/spack/spack/url.py", line 845, in substitute_version
    ver,  vs, vl, voffs) = substitution_offsets(path)
  File "/Users/Adam/spack/lib/spack/spack/url.py", line 789, in substitution_offsets
    ver,  vs, vl, vi, vregex = parse_version_offset(path)
  File "/Users/Adam/spack/lib/spack/spack/url.py", line 408, in parse_version_offset
    path, ext, suffix = split_url_extension(path)
  File "/Users/Adam/spack/lib/spack/spack/url.py", line 346, in split_url_extension
    match = re.search(r'(.*(?:sourceforge\.net|sf\.net)/.*)(/download)$', path)
  File "/Users/Adam/anaconda3/lib/python3.6/re.py", line 182, in search
    return _compile(pattern, flags).search(string)
```
It appears that Spack is trying to create the URL for the version during class construction, and since there is no global URL for that package, it crashes. There are currently 2 known workarounds:

1. Add a global URL attribute to the `package.py`
2. Add the version you want to use to the `package.py`

**Comment by adamjstewart** [comment]
Can you see if this is fixed by #3161?
